Relative source binding for gradient stop fix

kubaflo opened the pull request:
### Description of Change

Since the background property of the visual element control has never
had the parent set, the binding to {x:RelativeSource AncestorType} couldn't find
a proper control as the source for binding.

### Issues Fixed

Fixes https://github.com/dotnet/maui/issues/19253

https://github.com/dotnet/maui/assets/42434498/f3639e87-1b83-46c0-a3c8-85eb024d6c4c

kubaflo committed:
Added a Xaml test (#19253)

jsuarezruiz commented on src/Controls/src/Core/VisualElement/VisualElement.cs:
```diff
@@ -285,7 +285,10 @@ static void OnTransformChanged(BindableObject bindable, object oldValue, object
 				if (newvalue != null)
+				{
 					(bindable as VisualElement)?.NotifyBackgroundChanges();
+					((Brush)newvalue).Parent = bindable as VisualElement;
```
We are going to wait for the build and tests, but I think this will cause a memory leak.
cc @jonathanpeppers

jonathanpeppers replied:
Should something here at minimum _unset_ `Parent`?

kubaflo replied:
@jonathanpeppers I added an unsetting of a parent. Please let me know if it makes sense in this place.

jsuarezruiz commented:
/azp run

jsuarezruiz commented:
@kubaflo Could you rebase? Thanks!

kubaflo committed:
Fixed binding for GradientStop with a relative source (#19253)

Since the background property of the visual element control has never
had the parent set, the binding to {x:RelativeSource AncestorType} couldn't find
a proper control as the source for binding.

jsuarezruiz commented:
/azp run

jsuarezruiz commented:
/azp run

jsuarezruiz commented:
/azp run

kubaflo committed:
Fixes

mattleibow commented:
A question, what happens if you have 1 brush in a resource and you use it on 2 buttons?

I feel like only the last one will be notified of updates because you are switching parents.

I can understand if you have a view in the Resources, but for things like brush and image sources which are Element, this might break down.

@StephaneDelcroix is there a better mechanism for this, or is it the last wins for now.

StephaneDelcroix commented:
resources are sharing a single instance. we'd need x:Shared support for that

kubaflo commented:
> A question @kubaflo, what happens if you have 1 brush in a resource and you use it on 2 buttons?
> 
> I feel like only the last one will be notified of updates because you are switching parents.
> 
> I can understand if you have a view in the Resources, but for things like brush and image sources which are Element, this might break down.
> 
> @StephaneDelcroix is there a better mechanism for this, or is it the last wins for now.

Yes, with shared resources like this:

```
<ContentPage.Resources>
    <ResourceDictionary>
        <LinearGradientBrush x:Key="MyGradientBrush" StartPoint="0,0" EndPoint="1,0">
            <GradientStop Color="{Binding Source={RelativeSource AncestorType={x:Type Button}}, Path=BackgroundColor}"></GradientStop>
            <GradientStop Color="Green" Offset="1.0"/>
        </LinearGradientBrush>
    </ResourceDictionary>
</ContentPage.Resources>

<VerticalStackLayout Padding="20">
    <Button Text="Button 1" BackgroundColor="Blue" Background="{StaticResource MyGradientBrush}" />
    <Button Text="Button 2" BackgroundColor="Orange" Background="{StaticResource MyGradientBrush}" />
</VerticalStackLayout>
```

All the buttons will look like the last one unfortunately:
<img src="https://github.com/user-attachments/assets/76efc314-c76a-4d2e-8e72-b716ac4db215" width="300px"/>

kubaflo commented:
Closing since this issue can be avoided by using x: reference rather than relativeSource

```xaml
<Border
    BackgroundColor="Red"
    Stroke="Green"
    x:Name="border">
    <Border.Background>
        <LinearGradientBrush StartPoint="0,0"
                EndPoint="1,0">
            <GradientStop Color="{Binding BackgroundColor, Source={x:Reference border}}"
                    Offset="0.0"/>
            <GradientStop Color="{Binding Stroke.Color, Source={x:Reference border}}"
                    Offset="1.0"/>
        </LinearGradientBrush>
    </Border.Background>
</Border>
```
